src/rust/engine/logging/src/executor.rs

```diff
@@ -0,0 +1,88 @@
+use futures::Future;
+use std::sync::Arc;
+use tokio::runtime::Runtime;
+
+#[derive(Clone)]
+pub struct Executor {
+  runtime: Arc<Runtime>,
+}
+
+impl Executor {
+  pub fn new() -> Executor {
+    Executor {
+      runtime: Arc::new(
+        Runtime::new().unwrap_or_else(|e| panic!("Could not initialize Runtime: {:?}", e)),
+      ),
+    }
+  }
+
+  ///
+  /// Start running a Future on a tokio Runtime.
+  ///
+  pub fn spawn<F: Future<Item = (), Error = ()> + Send + 'static>(&self, future: F) {
+    // Make sure to copy our (thread-local) logging destination into the task.
+    // When a daemon thread kicks off a future, it should log like a daemon thread (and similarly
+    // for a user-facing thread).
+    let logging_destination = crate::get_destination();
+    self.runtime.executor().spawn(futures::lazy(move || {
+      crate::set_destination(logging_destination);
+      future
+    }))
+  }
+
+  ///
+  /// Run a Future on a tokio Runtime as a new Task. This is useful for blocking tokio tasks,
+  /// because blocking blocks the entire task called blocking, not just the part of it which is
+  /// marked as blocking. This allows finer granularity of marking tasks as blocking.
+  ///
+  /// See https://docs.rs/tokio-threadpool/0.1.15/tokio_threadpool/fn.blocking.html
+  ///
+  pub fn spawn_in_new_task<
+    Item: Send + 'static,
+    Error: Send + 'static,
+    F: Future<Item = Item, Error = Error> + Send + 'static,
+  >(
+    &self,
+    future: F,
+  ) -> impl Future<Item = Item, Error = Error> {
+    // Make sure to copy our (thread-local) logging destination into the task.
+    // When a daemon thread kicks off a future, it should log like a daemon thread (and similarly
+    // for a user-facing thread).
+    let logging_destination = crate::get_destination();
+    futures::sync::oneshot::spawn(
+      futures::lazy(move || {
+        crate::set_destination(logging_destination);
+        future
+      }),
+      &self.runtime.executor(),
+    )
+  }
+
+  ///
+  /// Run a Future and return its resolved Result.
+  ///
+  /// This should never be called from in a Future context.
+  ///
+  /// This method makes a new Runtime every time it runs, to ensure that the caller doesn't
+  /// accidentally deadlock by using this when a Future attempts to itself call Executor::spawn or
+  /// Executor::spawn_in_new_task. Because it should be used only in very limited situations, this
+  /// overhead is viewed to be acceptable.
+  ///
+  pub fn block_on<
```

Comment:
If we should "always" spawn when we use `block_on`, can this method be hidden behind that interface? ie, `spawn_and_block_on`?

Reply:
This doesn't do semantically different behaviour than before in `block_on`; in particular, this doesn't `spawn`, and none of the callers do `block_on(spawn(...))`.

This change:
1. Moves code that already existed in `Core`
2. Makes it use a new `Runtime` rather than the existing one, to cover up the fact that the code used to take a write lock on the Runtime, and that meant that if any of the work that was being blocked on _would_ try to `spawn`, it would fail deadlock waiting for the read lock on the `Runtime`.

Reply:
Aaaaah - I think the confusion here is because `blocking` and `block_on` are completely unrelated concepts. I will write some comments.